### PR TITLE
Feature/imp 1248/add mosaic carousel

### PR DESCRIPTION
**Pull request by SungjunApp**
ticket: https://pixlee.atlassian.net/browse/IMP-1248

## Background 
- Stradivarius needs a mosaic and horizontal list
- Sephora needs  a mosaic list

## Implementation

PXLWidgetView in production supports 
- Grid: displaying multiple items in a row
- List: displaying one item in a row

I additionally added 
- Mosaic: similar our Mosaic on Widget
- Horizontal: an horizontal version of List

## Demo
Please feel free to test it.
- app(APK) download link: https://drive.google.com/file/d/1J4Y69XaBzyN2JuyWjA1faquVpR1p-THd/view?usp=sharing

|Mosaic|Horizontal|
|---------|---------|
|||

#### Step 1: import this sample code into the app project
- layout
  ```xml
  <com.pixlee.pixleesdk.ui.widgets.list.PXLWidgetView
      android:id="@+id/widget"
      android:layout_width="match_parent"
      android:layout_height="match_parent"/>
  ```
- kotlin
  ```kotlin
  binding.widget.initiate(
      widgetTypeForAnalytics = "your_widget_type", // this will be used when this view automatically fires openedWidget, widgetVisible analytics
      viewType = PXLWidgetView.ViewType.Mosaic(gridSpan = 4, lineSpace = Space(lineWidthInPixel = 4.px.toInt(),includingEdge = false)),
      apiParameters = PXLKtxBaseAlbum.Params(
              // album images
              searchId = PXLKtxBaseAlbum.SearchId.Album(BuildConfig.PIXLEE_ALBUM_ID), // product images: searchId = PXLKtxBaseAlbum.SearchId.Product(BuildConfig.PIXLEE_SKU),
              filterOptions = PXLAlbumFilterOptions().apply {
                  // hasProduct and hasPermission are often used together for displaying photos with tagged products and gotten the permission from their creators
                  // if you don't see any photos after the loading is done, go to https://app.pixlee.com/app#albums/{your album id} and make sure your photos have the same filter conditions as your filterOptions.
                  hasProduct = true
                  hasPermission = true
  
                  // more filter options
                  // - hasPermission = true
                  // - inStockOnly = true
                  // - .. there are more. Please check README or PXLAlbumFilterOptions class for more filter options
              },
              sortOptions = PXLAlbumSortOptions().apply {
                  sortType = PXLAlbumSortType.RECENCY
                  descending = true
              }
      ),
      configuration = PXLPhotoView.Configuration().apply {
          pxlPhotoSize = PXLPhotoSize.MEDIUM
          imageScaleType = ImageScaleType.CENTER_CROP
      },
      loadMoreTextViewStyle = TextViewStyle().apply {
          text = "Load More"
          textPadding = TextPadding(0, 22.px.toInt(), 0, 22.px.toInt())
          size = 24.px
          color = Color.BLACK
      },
      onPhotoClickedListener = { view, photoWithImageScaleType ->
          // TODO: open lightbox
          ViewerActivity.launch(this, photoWithImageScaleType)
      }
  )
  ```
#### Step 2: Set up Horizontal or Mosaic
- To use Horizontal from PXLWidgetView, you can edit this and replace the value of viewType in the code example with your edited code.
  ```kotlin
  viewType = PXLWidgetView.ViewType.Horizontal(squareSizeInPixel = <your square Size In Pixel>, lineWidthInPixel = 4.px.toInt()),
  ```
- To use Mosaic from PXLWidgetView, you can edit this and replace the value of viewType in the code example with your edited code.
  ```kotlin
  viewType = PXLWidgetView.ViewType.Mosaic(gridSpan = 4, lineSpace = Space(lineWidthInPixel = 4.px.toInt(),includingEdge = false)),
  ```

### CODE REVIEW

File: pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/adapter/PXLPhotoAdapter.kt [hunk 1]
```diff
@@ -37,7 +45,7 @@ class PXLPhotoAdapter(
             is Item.Header -> TYPE_HEADER
             is Item.Content -> TYPE_ITEM
             is Item.LoadMore -> TYPE_LOAD_MORE
-
+            else -> TYPE_ITEM
```
Comment (jchen123) [inline pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/adapter/PXLPhotoAdapter.kt:48]: just sanity checking that this is correct syntax? i assume it is, ive just never seen it before ("is-else")

**Review by jchen123 — APPROVED** [review]
incredible

File: pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/adapter/PXLPhotoAdapter.kt (hunk 1 above)
Comment (jharrilim) [reply pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/adapter/PXLPhotoAdapter.kt:48]: [else is valid in Kotlin's pattern matching (`when`)](https://kotlinlang.org/docs/control-flow.html#when-expression), and can be used whether or not your cases use `is` or other conditions

File: simpleapp/src/main/res/layout/activity_dynamic_photos.xml [hunk 2]
```diff
@@ -14,11 +14,10 @@
         <com.pixlee.pixleesdk.ui.widgets.list.PXLWidgetView
```
Comment (jharrilim) [inline simpleapp/src/main/res/layout/activity_dynamic_photos.xml:14]: Is it possible to add an attribute here to specify which type of widget will be used or does that always have to happen in the Java/Kotlin?

**Review by jharrilim — APPROVED** [review]
Everything looks really good to me, I just left one question

File: pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/adapter/PXLPhotoAdapter.kt (hunk 1 above)
Comment (SungjunApp) [reply pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/adapter/PXLPhotoAdapter.kt:48]: yes that's possible in Kotlin which is same as this below. This's kotlin's official syntax guide https://kotlinlang.org/docs/typecasts.html#unsafe-cast-operator
```kotlin
val item = list[getRealPosition(position)]
        return if(item is Item.Header) TYPE_HEADER
        else if(item is Item.Content) TYPE_ITEM
        else if(item is Item.LoadMore) TYPE_LOAD_MORE
        else TYPE_ITEM
```

File: simpleapp/src/main/res/layout/activity_dynamic_photos.xml (hunk 2 above)
Comment (SungjunApp) [reply simpleapp/src/main/res/layout/activity_dynamic_photos.xml:14]: Yes, it is possible. As long as the attributes are intuitively easy to use, they can be good options.

**Pull request merged**